src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+                    {
+                        var variables = syntaxFacts.GetVariablesOfLocalDeclarationStatement(firstStatement);
+                        if (variables.Count == 1)
+                        {
+                            var firstVariable = (TVariableDeclaratorSyntax)variables[0];
+                            if (IsValidVariableDeclarator(firstVariable))
+                            {
+                                var firstVariableInitializer = syntaxFacts.GetValueOfEqualsValueClause(
+                                    syntaxFacts.GetInitializerOfVariableDeclarator(firstVariable));
+                                if (syntaxFacts.AreEquivalent(firstVariableInitializer, indexExpression))
+                                {
+                                    var type = (TTypeNode)syntaxFacts.GetTypeOfVariableDeclarator(firstVariable)?.WithoutLeadingTrivia();
+                                    var identifier = syntaxFacts.GetIdentifierOfVariableDeclarator(firstVariable);
+                                    var statement = firstStatement;
+                                    return (type, identifier, statement);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                return default;
+            }
+
+            void findAndReplaceMatches(SyntaxNode current)
+            {
+                if (syntaxFacts.AreEquivalent(current, indexExpression))
+                {
+                    // Found a match.  replace with iteration variable.
+                    var replacementToken = foreachIdentifierReference;
+
+                    if (semanticFacts.IsWrittenTo(semanticModel, current, cancellationToken))
+                    {
+                        replacementToken = replacementToken.WithAdditionalAnnotations(
+                            WarningAnnotation.Create(FeaturesResources.Warning_colon_Collection_was_modified_during_iteration));
+                    }
+
+                    if (crossesFunctionBoundary(current))
+                    {
+                        replacementToken = replacementToken.WithAdditionalAnnotations(
+                            WarningAnnotation.Create(FeaturesResources.Warning_colon_Iteration_variable_crossed_function_boundary));
+                    }
+
+                    var replacement = generator.IdentifierName(replacementToken).WithTriviaFrom(current);
+
+                    editor.ReplaceNode(current, replacement);
+                }
+
+                foreach (var child in current.ChildNodesAndTokens())
+                {
+                    if (child.IsNode)
+                    {
+                        findAndReplaceMatches(child.AsNode());
+                    }
+                }
+            }
+
+            bool crossesFunctionBoundary(SyntaxNode node)
+            {
+                var containingFunction = node.AncestorsAndSelf().FirstOrDefault(
```

Comment:
this looks like something data flow api also provide as well.

Reply:
You can get the set of captured variables.  And that's *almost* enough.  However, we want to put a warning at the actual capture point.  And, for that, we need to know if this is a location where a capture happened.

Reply:
you can First see whether it is captured, if not, no reason to walk down. 

if it is captured, then you can then walk down to find references.

also, using dataflow API is future proof since compiler determine the capture, not logic is duplicated. if compiler add yet another language feature that capture, this code is now broken. 

currently dataflow API doesn't give one locations but new Dataflow API (tagging @mavasani @AlekseyTs  @333fred @jinujoseph ) that based on IOperation I believe can provide reference locations.

I guess this is good candidate to use new IOperation data flow API once it is available.

Reply:
> I guess this is good candidate to use new IOperation data flow API once it is available.

Sure.  If we have that, we could consider using that here.  Until then, this approach works well.